feat(lua): add semver api

kelly-lin opened the pull request:
Fixes #20981.

Hi all, this is my first time contributing to Neovim so I apologise in advance if I did not format my code or this PR to the Neovim standard. If I have made any mistakes in the PR or the code, please point them out so that I can address them.

@justinmk this PR is complete with the features we discussed in the issues thread, can you please take a look?

Thanks.

kelly-lin committed:
feat(lua): add semver api

clason commented:
>  if I did not format my code or this PR to the Neovim standard.

That's easy :) Just run `stylua runtime/` from the repo root.

(And we use conventional commits, so it should be `feat(lua): add semver api` or similar.)

kelly-lin commented:
> > if I did not format my code or this PR to the Neovim standard.
> 
> That's easy :) Just run `stylua runtime/` from the repo root.
> 
> (And we use conventional commits, so it should be `feat(lua): add semver api` or similar.)
> 
> Oh, and you probably need to generate docs with `./scripts/gen_vimdoc.py` (needs a system installation of `luajit`).

Awesome thanks. I have amended the commit message.

I tried to run `./scripts/gen_vimdoc.py` but I got an error: "Requires 'nvim -l' feature, see https://github.com/neovim/neovim/pull/18706". I went to the link but it does not tell me how to get this feature. How do I get the `nvim -l` feature?

clason commented:
Oh, so that was changed. You need to use a current nightly, then -- the version in the repo no longer works with the current stable releases. @justinmk

kelly-lin commented:
> Oh, so that was changed. You need to use a current nightly, then -- the version in the master branch no longer works with the current stable releases. @justinmk
> 
> (you also need to add the new module to the list of files that documentation is generated for -- look at lines 134ff)

Thanks @clason, I ran `./scripts/gen_vimdoc.py` and `stylua runtime/` and got no changes. Should I expect `./scripts/gen_vimdoc.py` to generate some change to a file since I added in documentation in form of lua docstrings?

clason commented:
> Should I expect ./scripts/gen_vimdoc.py to generate some change to a file since I added in documentation in form of lua docstrings?

Yes, but you need to add your new file to the script first. (But this can wait till the tests are fixed ;))

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,248 @@
+  local v1_identifiers = vim.split(v1_parsed.prerelease, '%.')
+  local v2_identifiers = vim.split(v2_parsed.prerelease, '%.')
+  local i = 1
+  local max = vim.fn.max({ vim.tbl_count(v1_identifiers), vim.tbl_count(v2_identifiers) })
```
Better to use plain Lua `:h math.max()`

kelly-lin replied:
Awesome, thanks. Updated PR with changes.

kelly-lin replied:
As an aside, what is the reasoning behind using plain lua `math.max()` over `vim.fn.max`?

clason replied:
Why wouldn't you use a native function in your programming language rather than a Rube Goldberg machine that goes through three layers of indirection calling into a text editor to compare two numbers? ;)

clason replied:
(in other words, performance and robustness)

kelly-lin commented:
> > Should I expect ./scripts/gen_vimdoc.py to generate some change to a file since I added in documentation in form of lua docstrings?
> 
> Yes, but you need to add your new file to the script first. (But this can wait till the tests are fixed ;))

Ah I see, so I would need to add the module in `scripts/gen_vimdoc.py` in the  `CONFIG.lua.section_order` and `CONFIG.lua.files` for the doc generator to generate the docs?

> But this can wait till the tests are fixed ;)
Is this something that I have to do or is there an issue with the tests at the moment?

zeertzjq reviewed (commented):
Better to put the test in `test/functional/lua/` instead of `test/unit/`. Unittests aren't run with PUC Lua or on Windows, whereas functionaltests can always be run.

kelly-lin commented:
> Better to put the test in `test/functional/lua/` instead of `test/unit/`. Unittests aren't not run with PUC Lua or on Windows, whereas functionaltests can always be run.

Ah I see, I did not know that, thanks.

What is PUC Lua?

How would I know when to put a test in `test/unit/` over `test/functional/lua/`  and vice versa?

clason commented:
> What is PUC Lua?

It's the "Pontifical Catholic University" of Rio de Janeiro, where the original (hence PUC) Lua was developed -- as opposed to LuaJIT.

clason commented:
> How would I know when to put a test in test/unit/ over test/functional/lua/ and vice versa?

If in doubt, put it in `functional`, especially if it's testing a user-facing API.

zeertzjq commented:
> What is PUC Lua?

PUC Lua means Lua 5.1 without LuaJIT.

> How would I know when to put a test in `test/unit/` over `test/functional/lua/` and vice versa?

Unittests are intended to test C functions using FFI calls (which is only supported by LuaJIT). You are testing a Lua module, so you should put it in `test/functional/lua/`.

zeertzjq commented on test/functional/lua/semver_spec.lua:
```diff
@@ -0,0 +1,303 @@
+        { desc = 'build metadata before prerelease', version = '1.2.3+build.0-rc1' },
+      }
+      for _, tc in ipairs(testcases) do
+        it(string.format('returns error for %s: version = %s', tc.desc, tc.version), function()
```
```suggestion
        it(string.format('returns error for %s: version = %s', tc.desc, tostring(tc.version)), function()
```

kelly-lin replied:
Interesting, why is this test error happening on CI but not when I run locally using `TEST_FILE=test/functional/lua/semver_spec.lua make functionaltest`? Is there a command I can run to reproduce this locally?

zeertzjq replied:
Try changing `functionaltest` to `functionaltest-lua`

kelly-lin replied:
Still passing when I run `TEST_FILE=test/functional/lua/semver_spec.lua make functionaltest-lua`.

lewis6991 commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,248 @@
+    return 0
+  end
+
+  local v1_identifiers = vim.split(v1_parsed.prerelease, '%.')
```
```suggestion
  local v1_identifiers = vim.split(v1_parsed.prerelease, '.', { plain = true })
```

lewis6991 commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,248 @@
+---   that `semver.parse('v1.2)` *will throw an error*. When set to `false`
+---   `semver.parse('v1.2)` will coerce 'v1.2' to 'v1.2.0' and return the
+---   table: `{ major = 1, minor = 2, patch = 0 }`.
+--- @returns result number --- -1 if `v1 < v2`,0 if `v1 == v2` or, 1 if `v1 > v2`.
```
```suggestion
--- @return result number --- -1 if `v1 < v2`,0 if `v1 == v2` or, 1 if `v1 > v2`.
```

kelly-lin replied:
`@returns` (plural) is correct syntax, lua doc strings in other parts of the code base is also using `@returns` and the linter throws an error for `@return`.

lewis6991 replied:
I'm not sure this works with lua_lsp

kelly-lin replied:
I am using `sumneko_lua` and it is throwing a linting error when I try to use `@return` instead of `@returns`.

I did a quick `grep` through the code base and it looks like `@return` is only every used in `*.c, *.h` files. But all `*.lua` files use `@returns` and `@return` does not appear in any `*.lua` files.

lewis6991 replied:
Yes, this doesn't work with lua_lsp. You also need to drop `results` and change `number` to `integer`.

lewis6991 replied:
> I did a quick grep through the code base and it looks like @return is only ever used in *.c, *.h files. But all *.lua files use @returns and @return does not appear in any *.lua files.

Look in `treesitter.lua`.

lewis6991 commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,248 @@
+end
+
+---@private
+local function parse_prerelease(labels)
```
```suggestion
---@param labels string
local function parse_prerelease(labels)
```

clason commented:
As you're now iterating on type annotations, now would be a good time to add the new module to the docs like we discussed above.

(Also, a news entry is appropriate here, but that can wait until the end.)

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,266 @@
+---
+---  - `strict`: when set to `true` an error will be thrown for version
+---  strings which are not conforming to semver sepecifications, this means
+---  that `semver.parse('v1.2)` *will throw an error*. When set to `false`
```
Please do not use text markup in docstrings, as these will be translated to vimdoc, where `*` has a different meaning.

Also, this reads more like a documentation for `semver.parse`, not `cmp`. I think it's enough to mention that invalid version strings will throw an error and refer to `semver.parse` for details.

kelly-lin replied:
Thanks @clason, I have updated accordingly, please let me know if I did it correctly or not.

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,266 @@
+---
+---@param version string Version string formatted semantically.
+---@param opts table|nil Options
+---  Table containing options to configure parsing behaviour. The options and
+---  their default values are shown below:
```
Too verbose; take a look at other core modules (lsp, treesitter) for how option table keys are usually documented.

kelly-lin replied:
Updated this on the latest changes, can you please take a look and see if it is correct? Thanks.

max397574 commented on runtime/doc/lua.txt:
```diff
@@ -2473,4 +2473,47 @@ trust({opts})                                             *vim.secure.trust()*
+      • {version}  (string) Version string formatted semantically.
+      • {opts}     (table|nil) Options to configure parsing behaviour which
+                   takes the following optional keys:
+                   • strict (boolean): (default = false) when set to `true` an
```
as far as I remember at other places there is written `Defaults to true` and `bool` instead of `boolean` is used

kelly-lin replied:
Ah yes, you are right. I will amend.

max397574 replied:
I think we should wait for one of the core maintainers then
I just looked at a few helpages and just saw bool being used

lewis6991 replied:
Use `boolean`.

kelly-lin commented:
> As you're iterating on type annotations, now would be a good time to add the new module to the docs like we discussed above.
> 
> (Also, a news entry is appropriate here, but that can wait until the end.)

@clason I ran `scripts/gen_vimdoc.py` and added the new module to the script and have added the updated documentation in the PR. Can you please guide me to how I add in a news entry?

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+--- Compares the prerelease component of the two versions.
+---@param v1_parsed table Parsed version.
+---@param v2_parsed table Parsed version.
+---@return integer -1 if `v1_parsed < v2_parsed`,0 if `v1_parsed == v2_parsed` or, 1 if `v1_parsed > v2_parsed`.
```
```suggestion
---@return integer `-1` if `v1_parsed < v2_parsed`, `0` if `v1_parsed == v2_parsed`, `1` if `v1_parsed > v2_parsed`.
```

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+---@param opts table|nil Options to configure parsing behaviour with the optional keys:
+---                      - strict (boolean): (default = false) see `semver.parse` for
+---                      details.
+---@return integer -1 if `v1 < v2`,0 if `v1 == v2` or, 1 if `v1 > v2`.
```
```suggestion
---@return integer `-1` if `v1 < v2`, `0` if `v1 == v2`, `1` if `v1 > v2`.
```

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,262 @@
+---
+---@param version string Version string formatted semantically.
+---@param opts table|nil Options to configure parsing behaviour which takes the following optional keys:
+---                      - strict (boolean):  when set to `true` an error will be thrown for version
+---                      strings which are not conforming to semver sepecifications, this means that
+---                      `semver.parse('v1.2)` will throw an error. When set to `false`,
+---                      `semver.parse('v1.2)` will coerce 'v1.2' to 'v1.2.0' and return the table:
+---                      `{ major = 1, minor = 2, patch = 0 }`. Defaults to false.
```
```suggestion
---                      - strict (boolean): if `true`, raise an error if the version string does not conform to the specification; if `false` (default), parse missing fields as `0` (e.g., `semver.parse('v1.2)` will return `{ major = 1, minor = 2, patch = 0 }`.`
```

could (should) add a link to the specification?

clason reviewed (commented):
News entries go into `runtime/doc/news.txt`; this is a new feature, e.g.,

```
• Added a `vim.semver` module for parsing and comparing version numbers according to the semver specification, see |lua-semver|.

kelly-lin commented:
@clason I kept getting this error while trying to edit `news.txt`, any idea on how to fix?

![2023-02-20-235513](https://user-images.githubusercontent.com/19686599/220114687-1c33a62f-34b2-4f89-a00a-740f45a96bb7.jpg)

clason commented:
I suspect you are using nvim-treesitter. If you are, install (and make sure to keep updated) all parsers that are bundled with Neovim (c, lua, vim, help).

clason commented:
(and you always need to rerun the `gen_vimdoc` script anytime you change the docstrings...)

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+---
+---@param version string Version string formatted semantically.
+---@param opts table|nil Options to configure parsing behaviour which takes the following optional keys:
+---                      - strict (boolean):  when set to `true` an error will be thrown for version
+---                      strings which are not conforming to semver sepecifications (v2.0.0), see
+---                      https://semver.org/spec/v2.0.0.html for details. this means that
+---                      `semver.parse('v1.2)` will throw an error. When set to `false`,
+---                      `semver.parse('v1.2)` will coerce 'v1.2' to 'v1.2.0' and return the table:
+---                      `{ major = 1, minor = 2, patch = 0 }`. Defaults to false.
```
```suggestion
---                      - strict (boolean):  if `true`, throw an error for version
---                      strings that do not conform to semver specifications (                   https://semver.org/spec/v2.0.0.html), e.g.,
---                      `semver.parse('v1.2)`. If `false`,
---                      `semver.parse('v1.2)` will coerce 'v1.2' to 'v1.2.0' and return
---                      `{ major = 1, minor = 2, patch = 0 }`. Defaults to false.
```

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+--- Compares two strings (`v1` and `v2`) in semver format.
+---@param v1 string Version.
+---@param v2 string Version to be compared with v1.
+---@param opts table|nil Options to configure parsing behaviour with the optional keys:
```
```suggestion
---@param opts table|nil Optional keyword arguments:
```

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+--- `{ major = 1, minor = 0, patch = 1, prerelease = 'rc1', build = 'build.2' }`
+---
+---@param version string Version string formatted semantically.
+---@param opts table|nil Options to configure parsing behaviour which takes the following optional keys:
```
```suggestion
---@param opts table|nil Optional keyword arguments:
```

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+---
+--- `{ major = 1, minor = 0, patch = 1, prerelease = 'rc1', build = 'build.2' }`
+---
+---@param version string Version string formatted semantically.
```
```suggestion
---@param version string Version to be parsed.
```

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+
+--- Parses a semantically formatted version string into a table.
+---
+--- Supports leading "v" and leading and trailing whitespace in the version
+--- string, that means `"v1.0.1-rc1+build.2"`, `"1.0.1-rc1+build.2"`, `" v1.0.1-rc1+build.2"`
+--- and `"v1.0.1-rc1+build.2 "` will all produce the table:
+---
```
```suggestion
--- Supports leading "v" and leading and trailing whitespace in the version
--- string; e.g., `"v1.0.1-rc1+build.2"`, `"1.0.1-rc1+build.2"`, `" v1.0.1-rc1+build.2"`
--- and `"v1.0.1-rc1+build.2 "` will be parsed as
```

clason commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+---@param v2 string Version to be compared with v1.
+---@param opts table|nil Options to configure parsing behaviour with the optional keys:
+---                      - strict (boolean):  see `semver.parse` for details. Defaults to false.
+---@return integer `-1` if `v1 < v2`, `0` if `v1 == v2` or, `1` if `v1 > v2`.
```
```suggestion
---@return integer `-1` if `v1 < v2`, `0` if `v1 == v2`, `1` if `v1 > v2`.
```

kelly-lin commented:
@clason the formatting in runtime/doc/lua.txt does not look right, there is one line which is 335 characters long. How do I format the docstring so that it preserves the line breaks?

justinmk commented:
We already have `vim.version` but we could use metatable to add this surface area there. Can be done in a followup PR.

max397574 commented:
not sure if this is the right place to ask for this
but would it be possible to also include a `vim.semver.validate` function that just returns true or false if the string is a valid semver?
because as far as I understand there is an error if you pass an invalid version to `parse` 
this function would be useful for validating user input
it could also be extracted from/used in the `parse` function

kelly-lin commented:
> We already have `vim.version` but we could use metatable to add this surface area there. That avoids having both `vim.version` and `vim.semver`. All "version" stuff lives in `vim.version` then.
> 
> Can be done in a followup PR.

Would it be hard to do? I have not personally worked with metatables at all but happy to learn and make this change in this PR. Otherwise if we do it in a separate PR then it would mean introducing a breaking change later because `vim.semver` would not exist anymore.

justinmk commented:
> Otherwise if we do it in a separate PR then it would mean introducing a breaking change later because `vim.semver` would not exist anymore.

breaking change in the window of a few days is ok. If you want to try it as a separate commit here that's ok too.

kelly-lin commented:
> > Otherwise if we do it in a separate PR then it would mean introducing a breaking change later because `vim.semver` would not exist anymore.
> 
> breaking change in the window of a few days is ok. If you want to try it as a separate commit here that's ok too.

Maybe I will leave it out of this PR, I had a quick look and I am a bit lost to where to start, haha.

lewis6991 commented:
This is how you roughly do it:

```lua
local orig = vim.version -- whatever the function value is

vim.version = {} -- define vim.version as a table
vim.version.validate = semver.validate  -- add our new functions to the table
setmetable(vim.version, { __call = orig }) -- set a metatable that says to call the original vim.version function if this table is called like a function
```

justinmk reviewed (approved):
Looks like a great start! Let's do the `vim.version` rename in this PR since there isn't much else blocking it

justinmk commented on test/functional/lua/semver_spec.lua:
```diff
@@ -0,0 +1,306 @@
+          it(
+            string.format('returns correct table for %q: version = %q', tc.desc, tc.version),
+            function()
+              assert.same(tc.want, semver.parse(tc.version, { strict = true }))
```
use `eq()` unless there is a specific reason that assert.same is needed

kelly-lin replied:
Sorry I might not understand correctly. I tried changing to `assert.equals` here but the tests fail. This is the error
```
            Passed in:
            (table: 0x7fc046161308) {
              [major] = 10
              [minor] = 20
              [patch] = 123 }
            Expected:
            (table: 0x7fc046160168) {
              [major] = 10
              [minor] = 20
              [patch] = 123 }
```
am I misunderstanding something?

zeertzjq replied:
```lua
local helpers = require('test.functional.helpers')(after_each)
local eq = helpers.eq
```

justinmk commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+---@return table parsed_version Parsed version table
+function M.parse(version, opts)
+  if type(version) ~= 'string' then
+    error(create_err_msg(version))
```
error makes sense here, because this is like a "syntax" error (validation).

For other cases below, where the version string wasn't parseable for whatever reason, it seems useful to return `nil` instead of raising an error. Then the caller can use parse() as a "validate" function. That addresses this request:

> would it be possible to also include a vim.semver.validate function that just returns true or false if the string is a valid semver?

justinmk commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,263 @@
+
+  version = vim.trim(version)
+
+  local major, minor, patch, prerelease_and_build
+  if opts.strict then
+    major, minor, patch, prerelease_and_build = extract_components_strict(version)
+  else
+    major, minor, patch, prerelease_and_build = extract_components_loose(version)
+  end
```
ternary saves space and helps readability
```suggestion
  local major, minor, patch, prerelease_and_build = (opts.strict
    and extract_components_strict(version)
    or extract_components_loose(version))
```

kelly-lin replied:
When I try this, I get the linting error `The value is assigned as `nil` because the number of values is not enough. In Lua, `x, y = 1 ` is equivalent to `x, y = 1, nil` .`.

I tried this in a REPL and it looks like when we use this ternary form, all the return values except the first one gets dropped from the return.

```
> foo = function() return "foo", "bar" end
> foo()
foo     bar
> true and foo()
foo
> one, two = true and foo()
> one
foo
> two
nil
```

kelly-lin replied:
An alternative might be:
```
  local extract = opts.strict and extract_components_strict or extract_components_loose
  local major, minor, patch, prerelease_and_build = extract(version)
```
Thoughts?

kelly-lin commented:
> This is how you roughly do it:
> 
> ```lua
> -- get whatever the function value is
> local orig = vim.version 
> 
> -- define vim.version as a table
> vim.version = {}
> 
> -- add our new functions to the table
> vim.version.cmp = semver.cmp
> vim.version.parse = semver.parse
> 
> -- set a metatable that says to call the original vim.version function if this table is called like a function
> setmetatable(vim.version, { __call = orig }) 
> ```
> 
> `vim.version` is actually defined in `src/nvim/executor.c`. We should rename that to `_version`. And then define our new `vim.version` in `vim/_editor.lua`.

@lewis6991 I gave this a go. I made these changes on a separate branch, do you mind taking a look to see if I have done it correctly?
https://github.com/neovim/neovim/commit/d7df697bcb6545264bdd427e31b69810e7555bb0

justinmk commented:
> . I made these changes on a separate branch, do you mind taking a look to see if I have done it correctly? [c479775](https://github.com/neovim/neovim/commit/c47977597d350d40b777e31f6966558f0aff2236)

Looks fine. Style nit: use the same pattern as done for [vim.fn](https://github.com/neovim/neovim/blob/c47977597d350d40b777e31f6966558f0aff2236/runtime/lua/vim/_editor.lua#L256)

zeertzjq commented:
It seems that this module is currently only available via `require('vim.semver')`. Does the name of the file also need to be rename to `version.lua`, so it's available via `require('vim.version')` instead?

zeertzjq commented on test/functional/lua/semver_spec.lua:
```diff
@@ -0,0 +1,306 @@
+local semver = require('runtime.lua.vim.semver')
```
```suggestion
local semver = require('vim.semver')

local helpers = require('test.functional.helpers')(after_each)
local eq = helpers.eq
local pcall_err = helpers.pcall_err
```

zeertzjq commented on test/functional/lua/semver_spec.lua:
```diff
@@ -0,0 +1,306 @@
+      it(
+        string.format('returns %d if %s (v1 = %s, v2 = %s)', tc.want, tc.desc, tc.v1, tc.v2),
+        function()
+          assert.equals(tc.want, semver.cmp(tc.v1, tc.v2, { strict = true }))
```
```suggestion
          eq(tc.want, semver.cmp(tc.v1, tc.v2, { strict = true }))
```

zeertzjq commented on test/functional/lua/semver_spec.lua:
```diff
@@ -0,0 +1,306 @@
+          it(
+            string.format('returns correct table for %q: version = %q', tc.desc, tc.version),
+            function()
+              assert.same(tc.want, semver.parse(tc.version, { strict = true }))
```
```suggestion
              eq(tc.want, semver.parse(tc.version, { strict = true }))
```

zeertzjq commented on test/functional/lua/semver_spec.lua:
```diff
@@ -0,0 +1,306 @@
+          it(
+            string.format('returns correct table for %q: version = %q', tc.desc, tc.version),
+            function()
+              assert.same(tc.want, semver.parse(tc.version, { strict = false }))
```
```suggestion
              eq(tc.want, semver.parse(tc.version, { strict = false }))
```

zeertzjq commented on test/functional/lua/semver_spec.lua:
```diff
@@ -0,0 +1,306 @@
+        it(
+          string.format('returns error for %s: version = %s', tc.desc, tostring(tc.version)),
+          function()
+            assert.errors(function()
```
```suggestion
            pcall_err(function()
```

kelly-lin commented:
> It seems that this module is currently only available via `require('vim.semver')`. Does the name of the file also need to be rename to `version.lua`, so it's available via `require('vim.version')` instead?

Not sure about this. I just pushed some changes around exposing the module via `vim.version` using a metatable defined in `runtime/lua/vim/_editor`. Does this address what you are asking?

zeertzjq commented:
I'm not asking you to do anything. I'm asking whether @justinmk wants to file name of the module to be also renamed.

zeertzjq commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,259 @@
+--- string. e.g. `" v1.0.1-rc1+build.2"` , `"1.0.1-rc1+build.2"`, `"v1.0.1-rc1+build.2"`
+--- and `"v1.0.1-rc1+build.2 "` will be parsed as:
+---
+--- `{ major = 1, minor = 0, patch = 1, prerelease = 'rc1', build = 'build.2' }`
```
```suggestion
--- ``{ major = 1, minor = 0, patch = 1, prerelease = 'rc1', build = 'build.2' }``
```

zeertzjq commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,259 @@
+---                      - strict (boolean):  when set to `true` an error will be thrown for version
+---                      strings which are not conforming to semver sepecifications (v2.0.0), (see
+---                      semver.org/spec/v2.0.0.html for details), this means that
+---                      `semver.parse('v1.2)` will throw an error. When set to `false`,
+---                      `semver.parse('v1.2)` will coerce 'v1.2' to 'v1.2.0' and return the table:
```
```suggestion
---                      ``semver.parse('v1.2')`` will throw an error. When set to `false`,
---                      ``semver.parse('v1.2')`` will coerce "v1.2" to "v1.2.0" and return the table:
```

clason commented:
Yes, the filename should match that of the module; there is zero reason for them to be different.

noahfrederick commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,259 @@
+---@param version string Version string to be parsed.
+---@param opts table|nil Optional keyword arguments:
+---                      - strict (boolean):  when set to `true` an error will be thrown for version
+---                      strings which are not conforming to semver sepecifications (v2.0.0), (see
```
```suggestion
---                      strings that do not conform to the semver specification (v2.0.0), (see
```

justinmk commented on test/functional/lua/semver_spec.lua:
```diff
@@ -0,0 +1,310 @@
+local helpers = require('test.functional.helpers')(after_each)
```
rename file to `test/functional/lua/version_spec.lua`

justinmk commented on src/nvim/lua/executor.c:
```diff
@@ -741,7 +741,7 @@ static bool nlua_state_init(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
   // neovim version
   lua_pushcfunction(lstate, &nlua_nvim_version);
-  lua_setfield(lstate, -2, "version");
+  lua_setfield(lstate, -2, "_version");
```
if it makes things easier in this PR, go ahead and delete `nlua_nvim_version` and move its impl into `version.lua`. It's trivial:

```
return vim.fn.api_info().version
```

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -914,6 +914,9 @@ function vim._init_defaults()
 end
 
+local semver = require('runtime.lua.vim.semver')
+vim.version = setmetatable({ cmp = semver.cmp, parse = semver.parse }, { __call = vim._version })
```
```suggestion
vim.version = setmetatable(
  {
    cmp = semver.cmp,
    parse = semver.parse
  },
  {
    __call = function()
      return vim.fn.api_info().version
    end,
  })
```

justinmk commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,259 @@
+  }
+end
+
+return M
```
could do the setmetatable() stuff here, then we can name this file `version.lua`. Set the `__call` method to

```
   __call = function()
      return vim.fn.api_info().version
    end,
```

then we can delete the thing in executor.c

justinmk commented on runtime/lua/vim/semver.lua:
```diff
@@ -0,0 +1,259 @@
+  -- If major is nil then that means that the version does not begin with a
+  -- digit with or without a "v" prefix.
+  if major == nil then
+    error(create_err_msg(version))
```
return nil instead of raising an error. Then the caller can use parse() as a "validate" function. Caller can also opt-in to "error" behavior via doing `assert(vim.version(...))`.

lewis6991 commented on runtime/lua/vim/_editor.lua:
```diff
@@ -914,6 +914,13 @@ function vim._init_defaults()
+    return vim.fn.api_info().version
+  end,
+})
+
```
I think this can be moved to `_init_packages.lua`, around line 57

lewis6991 replied:
If you do it in verision.lua then you need to require in somewhere. By doing it in _init_packages, version.lua won't be loaded until vim.version is accessed, like all the other modules.

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2473,4 +2473,76 @@ trust({opts})                                             *vim.secure.trust()*
+        (integer) `-1` if `v1 < v2`, `0` if `v1 == v2`, `1` if `v1 > v2`.
+
+eq({version_1}, {version_2})                                    *version.eq()*
+    Compares `version_1` and `version_2` and returns `true` if the are equal. If `version_1` or `version_2` are not able to be parsed, an error will be thrown.
```
Too verbose. Similar for the gt/lt docstrings.

```suggestion
    Returns `true` if `version_1` are `version_2` are equal version strings.
```

clason replied:
Equal versions, but not necessary equal strings? (thinking of `v1.2` and `v1.2.0` with `strict=false`)

justinmk replied:
fixed

justinmk commented on runtime/doc/lua.txt:
```diff
@@ -2473,4 +2473,76 @@ trust({opts})                                             *vim.secure.trust()*
+        (boolean)
+
+gt({version_1}, {version_2})                                    *version.gt()*
+    Compares `version_1` and `version_2` and returns `true` if `version_1` > `version_2` . If `version_1` or `version_2` are not able to be parsed, an error will be thrown.
```
```suggestion
    Returns `true` if version string `v1` is greater than `v2` .
```

justinmk commented on test/functional/lua/version_spec.lua:
```diff
@@ -0,0 +1,570 @@
+      for _, tc in ipairs(testcases) do
+        it(string.format('for %s = %s', tc.version_1, tc.version_2), function()
+          matches(
+            'is not a valid version',
```
```suggestion
            'invalid version: foo',
```

zeertzjq replied:
I think you should comment on line 198 in version.lua

justinmk commented on runtime/lua/vim/version.lua:
```diff
@@ -0,0 +1,304 @@
+
+  -- If major is nil then that means that the version does not begin with a
+  -- digit with or without a "v" prefix.
+  if major == nil then
+    return nil
+  end
+
+  if
+    has_prerelease_and_build(prerelease_and_build)
+    and not is_prerelease_and_build_valid(prerelease_and_build)
+  then
+    return nil
+  end
```
`is_prerelease_and_build_valid` can return true if there is no prerelease/build.

```suggestion
  if major == nil or not is_prerelease_and_build_valid(prerelease_and_build)
  then
    return nil
  end
```

justinmk commented on runtime/lua/vim/version.lua:
```diff
@@ -0,0 +1,304 @@
+--- is not of any valid form, it will return false.
+---@param prerelease_and_build string
+---@return boolean
+local function is_prerelease_and_build_valid(prerelease_and_build)
```
why doesn' this use `parse_build` and `parse_prerelease` ?

lewis6991 commented on runtime/lua/vim/_init_packages.lua:
```diff
@@ -53,6 +53,13 @@ require('vim.shared')
 
 vim._submodules = { inspect = true }
 
+local version = require('vim.version')
+vim.version = setmetatable({ cmp = version.cmp, parse = version.parse }, {
+  __call = function()
+    return vim.fn.api_info().version
+  end,
+})
```
Integrate this into the block below. We don't want to `require'vim.version'` when we don't use it. The below metatable takes care of that.

```lua
if key == 'version' then
  local version = require('vim.version')    
  t[key] = setmetatable({ cmp = version.cmp, parse = version.parse }, {
  __call = function()
    return vim.fn.api_info().version
  end
  return t[key]
elseif ...
```

justinmk commented on runtime/lua/vim/version.lua:
```diff
@@ -0,0 +1,291 @@
+---@param version string
+---@return string
+local function create_err_msg(version)
+  return string.format('%s is not a valid version', version)
```
```suggestion
  return string.format('invalid version: "%s"', version)
```

bfredl commented on runtime/lua/vim/_init_packages.lua:
```diff
@@ -69,6 +69,14 @@ setmetatable(vim, {
       end
+    elseif key == 'version' then
+      local version = require('vim.version')
+      t[key] = setmetatable({ cmp = version.cmp, parse = version.parse }, {
+        __call = function()
+          return vim.fn.api_info().version
+        end,
+      })
```
We could use `setmetatable(M, { stuff })` inside `runtime/lua/vim/version.lua`, instead of hiding it here.

justinmk commented:
LGTM! Thanks for your patience and persistence @kelly-lin . 

LMK if you have any more changes. I will force-push a commit to this PR and then merge.

kelly-lin commented:
Thanks @justinmk @clason @lewis6991 @zeertzjq @bfredl for your help, patience and guidance. No more changes on my end. Ship it!

bfredl commented on runtime/lua/vim/_init_packages.lua:
```diff
@@ -69,6 +69,9 @@ setmetatable(vim, {
         return t[key]
       end
+    elseif key == 'version' then
+      t[key] = require('vim.version')
```
`version` can now be added to `vim._submodules` instead as it is no longer a special case.

justinmk committed:
refactor(vim.version): cleanup

- version.cmp(): assert valid version
- add test for loading vim.version (the other tests use shared.lua in
  the test runner)
- reduce test scopes, reword test descriptions

kelly-lin commented:
@justinmk I pushed an update to `news.txt` because I had errors in the help tags.

CI runners `test / ubuntu-22.04 functionaltest-lua (cc=gcc) (pull_request) ` and `test / ubuntu-22.04 functionaltest-lua (cc=gcc) (pull_request) ` keep failing but I have no idea why, any idea on how to fix?

justinmk commented:
just wait..

justinmk commented on runtime/lua/vim/version.lua:
```diff
@@ -0,0 +1,279 @@
+function M.cmp(v1, v2, opts)
+  opts = opts or { strict = false }
+  local v1_parsed = assert_version(v1, opts)
+  local v2_parsed = assert_version(v2, opts)
```
assert in cmp(),

- simplifies eq/gt/lt
- avoids passing nil to cmp_version_core

justinmk commented on test/functional/lua/version_spec.lua:
```diff
@@ -0,0 +1,571 @@
+describe('version', function()
+  it('package', function()
+    clear()
+    eq({ major = 42, minor = 3, patch = 99 }, exec_lua("return vim.version.parse('v42.3.99')"))
```
This tests the correctness of `runtime/lua/vim/_init_packages.lua` because `clear()` starts a new child Nvim process and `exec_lua` tries to use `vim.version`.

Previously, this went unnoticed because the rest of the tests use `require('vim.version')` (line 8) in the test runner process.

justinmk committed:
fix(vim.version): incorrect version.cmp()

Problem:
If major<major but minor>minor, cmp_version_core returns 1

Solution:
- Fix logic in cmp_version_core
- Delete most eq()/gt()/lt() tests, they are redundant.

justinmk commented:
- fixed a bug: [2113b5a](https://github.com/neovim/neovim/pull/22325/commits/2113b5ad97cce45e5d69e9f58880065a5bfabdd8)
- also deleted most of the eq/lt/gt tests, they are redundant and aren't worth the extra noise in the test logs.

justinmk commented:
parse() on Nvim's own version `v0.9.0-dev-1144+gca3bc56a3b1b` does not return the "build" component, is that expected?

```
{
    [major] = 1
    [minor] = 9
    [patch] = 3
    [prerelease] = 'dev'
    [build] = nil
}
```

Going to merge this anyway, can followup in another PR.

folke commented:
I've only seen this PR now, so I'm a bit late with my comment :)

I just wanted to share that lazy actually has a full semver implementation that you can find [here](https://github.com/folke/lazy.nvim/blob/main/lua/lazy/manage/semver.lua)

Apart from what has been merged, it also supports semver ranges and metamathods, that can be used like this:

```lua
version("0.9.0-dev+bui") > version("0.9.0")
-- or
range("~0.9").matches("0.9.0-dev+build")
range("1.2.3 - 2.3")
range("0.9.x").last(versions) --  returns the last version that matches the range
```

The ranges correctly process prereleases conforming with the spec.

You can find the semver tests [here](https://github.com/folke/lazy.nvim/blob/main/tests/manage/semver_spec.lua)

Feel free to pick whatever you want from the code if you think some of that functionality might be interesting to core.

justinmk commented:
> Feel free to pick whatever you want from the code if you think some of that functionality might be interesting to core.

Thank you @folke . Done in https://github.com/neovim/neovim/pull/22707
